File: src/test/java/dev/kensa/parse/CachingScenarioMethodAccessorTest.java
package dev.kensa.parse;

import dev.kensa.KensaException;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.Set;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class CachingScenarioMethodAccessorTest {

    private CachingScenarioMethodAccessor accessor;

    @BeforeEach
    void setUp() {
        accessor = new CachingScenarioMethodAccessor(new TestInstance(), Set.of("scenario"));
    }

    @Test
    void canAccessScenarios() {
        assertThat(accessor.valueOf("scenario", "aValue")).contains("FooBoo!!");
        assertThat(accessor.valueOf("scenario", "aNullValue")).isEmpty();
        assertThat(accessor.valueOf("nullScenario", "aValue")).isEmpty();
        assertThat(accessor.valueOf("scenarioDoesNotExist", "foo")).isEmpty();
        assertThatThrownBy(() ->accessor.valueOf("scenario", "methodDoesNotExist")).isInstanceOf(KensaException.class);
    }

    @Test
    void canCacheReturnValues() {
        assertThat(accessor.valueOf("scenario", "aValue")).contains("FooBoo!!");
        assertThat(accessor.valueOf("scenario", "aValue")).contains("FooBoo!!");
    }

    private static class TestInstance {
        private TestScenario scenario = new TestScenario();
        private TestScenario nullScenario = null;
    }

    private static class TestScenario {
        private String value = "FooBoo!!";

        public String aValue() {
            String value = this.value;
            this.value = "Meh!";
            return value;
        }
        public String aNullValue() {
            return null;
        }
    }
}

File: src/test/java/dev/kensa/parse/CachingScenarioMethodAccessorTest.kt
package dev.kensa.parse

import dev.kensa.parse.Accessor.ValueAccessor.PropertyAccessor
import dev.kensa.util.allProperties
import io.kotest.matchers.nulls.shouldBeNull
import io.kotest.matchers.shouldBe
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

internal class CachingScenarioMethodAccessorTest {

    private lateinit var accessor: CachingScenarioMethodAccessor

    @BeforeEach
    internal fun setUp() {
        accessor = CachingScenarioMethodAccessor(JavaTestInstance(),
            JavaTestInstance::class.allProperties.map { PropertyAccessor(it) }.associateBy { it.name })
    }

    @Test
    internal fun `can access scenarios`() {
        accessor.valueOf("scenario", "aValue") shouldBe "FooBoo!!"
        accessor.valueOf("scenario", "aNullValue").shouldBeNull()
        accessor.valueOf("nullScenario", "aValue").shouldBeNull()
        accessor.valueOf("scenarioDoesNotExist", "foo").shouldBeNull()
    }

    @Test
    internal fun `caches return values`() {
        accessor.valueOf("scenario", "aValue") shouldBe "FooBoo!!"
        accessor.valueOf("scenario", "aValue") shouldBe "FooBoo!!"
    }
}